src/transformers/models/deepseek_v3/modeling_deepseek_v3.py

```diff
@@ -209,116 +185,84 @@ class DeepseekV3MoE(nn.Module):
-        idxs = topk_ids.view(-1).argsort()
-        sorted_tokens = x[idxs // topk_ids.shape[1]]
-        sorted_tokens_shape = sorted_tokens.shape
-        if self.ep_size > 1:
-            tokens_per_ep_rank = tokens_per_expert.view(self.ep_size, -1).sum(dim=1)
-            tokens_per_expert_group = tokens_per_expert.new_empty(tokens_per_expert.shape[0])
-            dist.all_to_all_single(tokens_per_expert_group, tokens_per_expert)
-            output_splits = tokens_per_expert_group.view(self.ep_size, -1).sum(1).cpu().numpy().tolist()
-            gathered_tokens = sorted_tokens.new_empty(
-                tokens_per_expert_group.sum(dim=0).cpu().item(), sorted_tokens.shape[1]
-            )
-            input_split_sizes = tokens_per_ep_rank.cpu().numpy().tolist()
-            dist.all_to_all(
-                list(gathered_tokens.split(output_splits)),
-                list(sorted_tokens.split(input_split_sizes)),
-            )
-            tokens_per_expert_post_gather = tokens_per_expert_group.view(self.ep_size, self.experts_per_rank).sum(
-                dim=0
-            )
-            gatherd_idxs = np.zeros(shape=(gathered_tokens.shape[0],), dtype=np.int32)
-            s = 0
-            for i, k in enumerate(tokens_per_expert_group.cpu().numpy()):
-                gatherd_idxs[s : s + k] = i % self.experts_per_rank
-                s += k
-            gatherd_idxs = gatherd_idxs.argsort()
-            sorted_tokens = gathered_tokens[gatherd_idxs]
-            tokens_per_expert = tokens_per_expert_post_gather
+    def moe_infer(self, hidden_states: torch.Tensor, topk_indices: torch.Tensor, topk_weights: torch.Tensor):
+        """
+        Perform inference using a Mixture of Experts (MoE) model.
+
+        Args:
+            hidden_states (torch.Tensor): Input hidden states.
+            topk_indices (torch.Tensor): Indices of the top-k experts for each token.
+            topk_weights (torch.Tensor): Weights associated with the top-k experts.
+
+        Returns:
+            torch.Tensor: Output of the MoE model.
+        """
+        num_experts = len(self.experts)
+        batch_size, num_topk = topk_indices.shape
+
+        # Count the number of tokens assigned to each expert
+        expert_counts = topk_indices.new_zeros((batch_size, num_experts))
+        expert_counts.scatter_(1, topk_indices, 1)
+        tokens_per_expert = expert_counts.sum(dim=0)
+
+        # Sort tokens by their assigned expert
+        sorted_indices = topk_indices.view(-1).argsort()
+        sorted_tokens = hidden_states[sorted_indices // num_topk]
         tokens_per_expert = tokens_per_expert.cpu().numpy()
 
-        outputs = []
-        start_idx = 0
-        for i, num_tokens in enumerate(tokens_per_expert):
-            end_idx = start_idx + num_tokens
+        # Process tokens through their assigned experts
+        expert_outputs = []
+        current_pos = 0
+
+        for expert_idx, num_tokens in enumerate(tokens_per_expert):
```

Comment:
nice, IDK how many experts are actually there, but given how important the model is, we might want to use more optimized version of MoE implementation. The best we have in `transformers` is for SwitchTransformers ! Were we skip experts that won't have any tokens. 

One thing is also that this is not compile compatible "yet"?

Reply:
I will check this out! thanks :)